wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs

```diff
@@ -0,0 +1,194 @@
+newLegacyKeystore ctx = do
+     us <- atomically $ readTVar $ ncUserSecret ctx
+     Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+-- | Creates a 'Keystore' out of a randomly generated temporary file (i.e.
+-- inside your $TMPDIR of choice). Suitable for testing.
+newInMemoryTransientKeystore :: MonadIO m => m Keystore
+newInMemoryTransientKeystore = liftIO $ runIdentityT $ fromKeystore $ do
+    tempDir         <- liftIO getTemporaryDirectory
+    (tempFile, hdl) <- liftIO $ openTempFile tempDir "keystore.key"
+    liftIO $ hClose hdl
+    us <- takeUserSecret tempFile
+    Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+
+-- | Release the resources associated with this 'Keystore'.
+-- This function is idempotent and can be called multiple times.
+releaseKeystore :: MonadIO m => Keystore -> m ()
+releaseKeystore (Keystore ks) = liftIO $ modifyMVar ks $ \(InternalStorage us st) ->
+    case st of
+         StorageReleased      -> return ((InternalStorage us StorageReleased), ())
+         StorageInitialised   -> do
+             writeUserSecretRelease us
+             return ((InternalStorage us StorageReleased), ())
+
+
+{-------------------------------------------------------------------------------
+  Destroying a keystore
+-------------------------------------------------------------------------------}
+
+-- | Destroys a 'Keystore'.
+-- Completely obliterate the keystore from disk, with all its secrets.
+-- This operation cannot be reverted.
+-- This is a very destructive option that most of the time you
+-- probably don't want, as it will effectively leave you with an empty
+-- 'MVar' and your next call to an operation on the keystore would block
+-- indefinitely.
+-- This has still its use in some teardowns or tests.
+-- Note that this operation will always succeed. Use with care.
+destroyKeystore :: MonadIO m => Keystore -> m ()
+destroyKeystore (Keystore ks) = liftIO $ do
+    (InternalStorage us _st) <- takeMVar ks
+    writeUserSecretRelease us
+    removeFile (getUSPath us)
+
+{-------------------------------------------------------------------------------
+  Inserting things inside a keystore
+-------------------------------------------------------------------------------}
+
+-- | Insert a new 'EncryptedSecretKey' indexed by the input 'WalletId'.
+insert :: MonadIO m
+       => WalletId
+       -> EncryptedSecretKey
+       -> Keystore
+       -> m ()
+insert _walletId esk (Keystore ks) =
+    liftIO $ modifyMVar_ ks $ \(InternalStorage us st) -> do
+        return $ if view usKeys us `contains` esk
+                     then InternalStorage us st
+                     else InternalStorage (us & over usKeys (esk :)) st
+    where
+      -- Comparator taken from the old code which needs to hash
+      -- all the 'EncryptedSecretKey' in order to compare them.
+      contains :: [EncryptedSecretKey] -> EncryptedSecretKey -> Bool
+      contains ls k = hash k `elem` map hash ls
```

Comment:
Not your fault, but let's add a TODO (or open a ticket): keeping all keys in memory until the node gets shut down is madness. If you lose power or something, you lose all data. This needs to be fixed.